src/Runtime/jni/src/com/ibm/onnxmlir/OMModel.java

```diff
@@ -1,58 +1,81 @@
-                                        .getLocation().toURI());
-            jarDir = jar.getParentFile().getAbsolutePath();
-            libPath = jarDir + "/" + libname;
-
-            // Open jar file to read and check libname inside jar.
-            // If IOException thrown, load .so from where .jar is.
-            //
-            // Checking whether DynEntryPoint.class.getResourceAsStream returns null
-            // does NOT work. Because it checks whether the resource is
-            // available on the classpath, not only just inside the jar file.
-            jf = new JarFile(jar);
-            if (jf.getEntry(libname) != null) {
-                File libFile = new File(libPath);
-                // Copy .so to where jar is
-                Files.copy(jf.getInputStream(jf.getEntry(libname)),
-                        libFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
-                // z/OS USS requires "x" permission bit
-                libFile.setExecutable(true, false);
-                // Load the temporary .so copy
-                System.load(libPath);
-                // POSIX can unlink file after loading
-                libFile.delete();
-            }
-            else {
-                // Throw subclass of IOException
-                throw new FileNotFoundException(".so not found inside jar");
-            }
-        } catch (URISyntaxException e) {
-            // Failed to find jar path, assume the .so is in cwd
-            System.load(libname);
-        } catch (IOException e) {
-            // .so not found in jar, assume it's where jar is
-            System.load(libPath);
+            // Get path of jar
+	    Path jar = Paths.get(OMModel.class.getProtectionDomain()
+                                              .getCodeSource()
+                                              .getLocation().toURI());
+	    Path jarDir = jar.getParent();
+
+	    // All .so files found will be copied under the ./native subdirectory
+	    // where the jar file resides.
+	    Path tmpDir = Files.createTempDirectory(jarDir, "native_");
+
+            // Open jar file to load all .so libs inside. The .so is loaded
+            // by OS so it must be extracted to the file system.
+            JarFile jf = new JarFile(jar.toFile());
+	    for (Enumeration<JarEntry> e = jf.entries(); e.hasMoreElements(); ) {
+		String libFile = e.nextElement().getName();
+
+		if (libFile.endsWith(".so")) {
+		    try {
+			Path lib = Paths.get(tmpDir.toString(), libFile);
+			Path libDir = lib.getParent();
+
+			Files.createDirectories(libDir);
+
+			// Copy .so to the temporary directory
+			Files.copy(jf.getInputStream(jf.getEntry(libFile)), lib,
+				   StandardCopyOption.REPLACE_EXISTING);
+
+			// z/OS USS requires "x" permission bit
+			Files.setPosixFilePermissions(lib,
+			      PosixFilePermissions.fromString("rwxr-xr-x"));
+
+			// Load the temporary .so copy
+			System.load(lib.toString());
```

Comment:
Is there any concept of ordering the loading of libraries?  For example if you want library foo.so loaded before bar.so?  Right now you seem to be at the mercy of the order of the files in the jar.

Reply:
Nothing is being resolved here. The libraries are simply being loaded to make the symbols available so it makes no difference which library is loaded first.